Transform into the correct embed block based on URL patterns

notnownikki opened the pull request:
## Description

First stage in decoupling from the oEmbed metadata. This addresses provider names, and makes sure that URLs get handled by the correct embed block. This has the nice effect that we can have custom settings per embed provider, and the user won't miss out on them if they select the wrong block, or the generic embed block.

## How Has This Been Tested?

Insert a generic embed block. Make sure the block settings are visible. Enter a Twitter url, click embed. You should see the block settings change to reflect that it's become a Twitter block.

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,32 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
 const HOSTS_NO_PREVIEWS = [ 'facebook.com' ];
 
-function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, keywords = [] } ) {
+const wpEmbedAPI = memoize( ( url ) => wp.apiRequest( { path: `/oembed/1.0/proxy?${ stringify( { url } ) }` } ) );
```
Cache the embed API responses, so if things get transformed, or deleted and added again, we're not spamming the API.

aduth replied:
>Cache the embed API responses, so if things get transformed, or deleted and added again, we're not spamming the API.

Might be valuable to add as an inline code comment?

notnownikki replied:
Added!

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -286,6 +337,7 @@ export const common = [
 			title: 'Twitter',
 			icon: 'embed-post',
 			keywords: [ __( 'tweet' ) ],
+			patterns: [ '(www\.)?twitter\.com\/.+' ],
```
Why do we have regexes here? Why can't we discover these from the embed API?

The provider name comes from the oembed response, so if you look at the list of providers in `wp-includes/class-oembed.php` you'll see it's just an array of regexes and oembed URLs. There's nothing there to tie each provider to a block.

So these patterns are used to match a URL to a block. They're not the full regex used by oembed, they just match enough of the URL to be able to match it to the block.

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -110,20 +140,41 @@ function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, k
+							// Some plugins put the embed html in `result`, so get the right one here.
+							const html = obj.html ? obj.html : obj.result;
+							// Block titles are the provider names, but we can override with `providerName` too.
+							const providerNameSlug = kebabCase( toLower( '' !== providerName ? providerName : title ) );
```
Why don't we take the providerName from the oembed response any more?

Because there are plugins that bypass oembed completely, and don't provide it. And the titles of the blocks are the provider names anyway.

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -110,20 +140,41 @@ function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, k
+							const html = obj.html ? obj.html : obj.result;
+							// Block titles are the provider names, but we can override with `providerName` too.
+							const providerNameSlug = kebabCase( toLower( '' !== providerName ? providerName : title ) );
+							let { type = 'rich' } = obj;
```
Again, some plugins don't return a type, so default this to `rich`, as those plugins are returning HTML content and not metadata about other objects (like oembed does about photos)

aduth replied:
Sounds like a good inline code comment.

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -515,13 +596,15 @@ export const others = [
 		settings: getEmbedBlockSettings( {
 			title: 'WordPress.tv',
 			icon: 'embed-video',
+			patterns: [ 'wordpress\.tv\/.+' ],
```
Yes, I put all these in by hand. Yes, I'm comfortable with that. I am enlightened. Thanks for reading this far!

mkaz reviewed (approved):
I've tested this out with various embeds and scenarios converting back and forth. On load, on paste, and the wrong embed in the wrong block and this auto converts them all quite nicely.

aduth commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,33 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+// Caches the embed API calls, so if blocks get transformed, or deleted and added again, we don't spam the API.
+const wpEmbedAPI = memoize( ( url ) => wp.apiRequest( { path: `/oembed/1.0/proxy?${ stringify( { url } ) }` } ) );
+
+const matchesPatterns = ( url, patterns ) => {
```
This pattern of logic is what [`Array#some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) is perfectly suited for:

```js
return patterns.some( ( pattern ) => (
	url.match( new RegExp( '^https?:\/\/' + patterns[ patternIndex ] + '$', 'i' ) )
) );
```

aduth commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,33 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+
+const matchesPatterns = ( url, patterns ) => {
+	for ( const patternIndex in patterns ) {
+		const pattern = new RegExp( '^https?:\/\/' + patterns[ patternIndex ] + '$', 'i' );
```
Rather than constructing these regular expressions, should this just be duplicated into all of the patterns below? Not sure it's obvious to an implementer that they need to: provide string which is a regular expression, but not including protocol, or end anchor, or flags.

What about providers which explicitly don't support HTTPS? 

https://github.com/WordPress/wordpress-develop/blob/3a46cba/src/wp-includes/class-oembed.php#L63-L64

I can appreciate the effort to DRY this, but to me it causes more confusion to the maintainer than it helps.

aduth commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,33 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+};
+
+const findBlock = ( url ) => {
+	const findMatch = ( blocks ) => {
```
Why do we define a function here? Couldn't we just assign `const blocks = [ ...common, ...others ]` and collapse the function into the outer function?

aduth commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,33 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+
+const findBlock = ( url ) => {
+	const findMatch = ( blocks ) => {
+		for ( const blockIndex in blocks ) {
```
`for...in` should generally not be used for array iteration:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#Array_iteration_and_for...in

Further, the performance characteristics between the two largely favor a standard for loop:

https://jsperf.com/for-loop-vs-for-in-loop

For what you're doing here, I'd suggest a standard `for ( let i = 0; ...`.

Or a similar effect could be achieved with [`_.find`](https://lodash.com/docs/4.17.5#find) (there is a native [`Array#find`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) but it is **not** polyfilled in Gutenberg):

```js
function findBlock( url ) {
	const blocks = [ ...common, ...others ];
	const block = find( blocks, ( block ) => {
		return matchesPatterns( url, block.settings.patterns );
	} );

	if ( ! block ) {
		return 'core/embed';
	}

	return block.name;
}
```

aduth commented on blocks/library/embed/index.js:
```diff
@@ -38,6 +65,10 @@ function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, k
 
 		keywords,
 
+		providerName,
```
These are not properties of the block API, which we are registering here. It actually makes me think we ought to be warning on non-recognized properties on the registered block instance.

Instead, you could just move `findBlock` into the scope of this function so it has direct access to `patterns`.

aduth replied:
Where is `providerName` assigned? If it is assigned, should it be assigned as the default for `providerNameSlug`? Might be useful to avoid block unneeded block updates for known block types.

notnownikki replied:
`findBlock` needs access to the patterns from all of the embed blocks though, because when the user tries to embed a URL and that URL doesn't match the patterns of the current block, it has to go through all the other embed block settings trying to find one it matches, so I'm not sure moving it into that scope would help? Maybe I'm missing something here though.

notnownikki replied:
`providerName` comes from the embed API. The current block titles are based on the provider names, and as there are plugins which do not provide a provider name, I've updated this so that it's not part of the block settings and if we get it from the API we use it, if not we use the title.

aduth replied:
>`findBlock` needs access to the patterns from all of the embed blocks though, because when the user tries to embed a URL and that URL doesn't match the patterns of the current block, it has to go through all the other embed block settings trying to find one it matches,

Does it? `getEmbedBlockSettinsg` is used to create many individual instances of blocks. Each block should only be responsible for transforming into itself, not other blocks. A block should have no awareness of other blocks existing, even here for embeds.

aduth commented on blocks/library/embed/index.js:
```diff
@@ -110,20 +141,41 @@ function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, k
 								return;
 							}
-
-							const { html, provider_name: providerName } = obj;
```
I'm confused. I thought the reason we needed to manually add all block regular expressions was because we had no means on the server to recognize the provider name. So where does this value originate?

notnownikki replied:
We have the means _sometimes_. The oEmbed API does provide this data, but in the case where the user has a plugin that handles YouTube embeds (for example) and bypasses the oEmbed code, we don't get any of the metadata, so we don't get the provider name. I guess we could handle this like `type`, and take it if it's there?

notnownikki replied:
So i'm going to remove this from the block settings, and use the block's title if we don't get a provider name from the embed API.

aduth commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,33 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
 
-function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, keywords = [] } ) {
+// Caches the embed API calls, so if blocks get transformed, or deleted and added again, we don't spam the API.
+const wpEmbedAPI = memoize( ( url ) => wp.apiRequest( { path: `/oembed/1.0/proxy?${ stringify( { url } ) }` } ) );
```
Once #5219 lands, we're going to want to update this block to not depend on `wp.apiRequest` directly.

notnownikki replied:
I'd like to update this to use `withSelect` in a separate PR as it's a big change to the edit component, and I'd rather keep this PR as small as possible. Would this PR be ok to go in as it is? I've already started removing the direct `wp.apiRequest` call, but for ease of review I'd rather it was another PR.

notnownikki commented:
@aduth review points should be addressed now

aduth commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,27 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+
+const matchesPatterns = ( url, patterns ) => {
+	return patterns.some( ( pattern ) => {
+		return url.match( new RegExp( pattern, 'i' ) );
```
I don't think we need to construct this from a string any longer, just define the original values as RegExp in the first place.

e.g.

**Before:**

```js
patterns: [ '^https?:\/\/(www\.)?speakerdeck\.com\/.+' ],
```

**After:**

```js
patterns: [ /^https?:\/\/(www\.)?speakerdeck\.com\/.+/i ],
```

notnownikki replied:
Done!

aduth commented on blocks/library/embed/index.js:
```diff
@@ -38,6 +59,8 @@ function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, k
 
 		keywords,
 
+		patterns,
```
If it's not a supported property of the block API, it shouldn't be assigned.

notnownikki commented:
... or at least query if 'core/paragraph' exists to see if we have a path forward.

notnownikki commented:
@youknowriad @iseulde @mtias could I get your feedback on this too please? The "blocks have no knowledge of each other" rule leaves me without a solution here.

Summary:

This is the start of work to solve two problems, and provide an enhancment.

1) https://github.com/WordPress/gutenberg/issues/4799 - pasting an unsupported URL into a paragraph block leaves you with an embed block that says the content can't be embedded. It should just give you the URL in a pargraph block with the URL in it, but if the embed block can't transform itself back into a paragraph block if the response from the embed API shows that it can't be embedded, how do we solve this?

2) Making sure the user gets the correct embed block for the URL. Blocks like Twitter can have extra options, and we want to make sure the user gets the correct embed block for the URL they use. If a user puts a Twitter URL into a YouTube embed block, it should just switch over seamlessly. If we can't query something to see if there's a block to handle the URL, what can we do here?

3) Creating URL preview blocks with OpenGraph data (or content scraped from the URL) in the same way that Facebook or G+ do. This work was started in https://github.com/WordPress/gutenberg/pull/5075 , but has been closed while we get this PR sorted out, and right now it's blocked on the "no knowledge of other blocks" rule. If we can't make new blocks or transform the current block, I don't see a way to do this, because we have to make a decision on what to do based on what the embed API gives back.

notnownikki commented:
As far as not having the patterns as part of the block settings, we could keep a registry of patterns -> block inside the module, and query them instead of looping through block settings, that seems cleaner than dumping them into settings.

aduth commented:
I'm fine to scale back how strict we are about unawareness of other blocks, particularly in how this file itself is registering all embed blocks. But this should probably take the form of a top-level function, and not add unsupported properties to the block object.

aduth commented:
>#4799 - pasting an unsupported URL into a paragraph block leaves you with an embed block that says the content can't be embedded. It should just give you the URL in a pargraph block with the URL in it, but if the embed block can't transform itself back into a paragraph block if the response from the embed API shows that it can't be embedded, how do we solve this?

I think we need to stop being so aggressive with our transforms, and only transform _known_ block types. Things like `core-embed/wordpress` that can't be known by the URL alone? Just ignore them. If the user really seeks to embed, they can simply add an embed block.

A potential future enhancement can be _trying_ to get an embed result, and only transforming after it's been confirmed that it's a valid embeddable URL.

notnownikki commented:
> I think we need to stop being so aggressive with our transforms, and only transform known block types. Things like core-embed/wordpress that can't be known by the URL alone? Just ignore them. If the user really seeks to embed, they can simply add an embed block.

I half agree, but the half that doesn't agree is saying that this results in a worse experience than the existing editor, where you can paste in a URL and have it embed.

> A potential future enhancement can be trying to get an embed result, and only transforming after it's been confirmed that it's a valid embeddable URL.

Depending on how we do that, it's either easy, or horrible. The problem there is that checking if the URL is embeddable is asynchronous, and our transform code doesn't deal with resolving promises, and making it do that seems to be a whole other rabbit hole of nastiness.

The other way is to create the embed block, and transform back into a paragraph if it's not embeddable, and to the correct embed block if it is. That's pretty much what's happening in this PR.

There's an option that seems to work here but in practise doesn't - that's having all the URL matching patterns in JS, so we know for sure if a URL can be embedded before we transform to a block. However, because plugins can add / remove / alter embed providers, we don't know if our blocks match what oembed has, and there's no API to discover providers and patterns. That's why the patterns here just match the domains of the sites.

notnownikki commented:
Rebased on master, and I moved the patterns out of the settings so we don't store non-standard fields. Instead, we map block names to settings when we get a new embed block definition, so it's a little cleaner than before.

I'm still looking at the `wp.data` module to see if we can move `wp.apiRequest` out of the server side render method, but it looks like it would make the code much more complex and less readable, so I'm not sure it's a suitable use case for that yet.

aduth reviewed (commented):
Apologies for the delay in circling back to this. It works quite well. Left a few remarks that we can discuss, though I'd not want to hold this up much longer.

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,28 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+// A map of block names to URL patterns, so we can find which block should handle a URL.
+const blockPatterns = {};
+
+const matchesPatterns = ( url, patterns ) => {
```
I'm learning Lisp, and... parentheses... parentheses everywhere... I guess it spilled over :)

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -110,20 +132,45 @@ function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, k
-							const providerNameSlug = kebabCase( toLower( providerName ) );
-							let { type } = obj;
-
+							// Some plugins put the embed html in `result`, so get the right one here.
```
Yeah, agreed. I'll remove this, and it can be handled much better one the data PR lands that moves the apiRequest out of the block.

notnownikki replied:
Removed!

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,28 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+const wpEmbedAPI = memoize( ( url ) => wp.apiRequest( { path: `/oembed/1.0/proxy?${ stringify( { url } ) }` } ) );
+
+// A map of block names to URL patterns, so we can find which block should handle a URL.
+const blockPatterns = {};
```
This was done to keep unsupported properties out of the block settings, in response to https://github.com/WordPress/gutenberg/pull/5315#discussion_r173516040

notnownikki replied:
Also, if plugins want to register embed blocks at some point in the future, this is a step to support that.

aduth replied:
Yes, but we can still iterate `common` and `others` without including them in the block definition. They're available in the top-level scope.

notnownikki replied:
I must be missing something here :) If I iterate over common and others, how would I access the patterns? They're a bunch of block settings, and we don't want unsupported properties in them. So what could we do?

aduth replied:
>They're a bunch of block settings, and we don't want unsupported properties in them. So what could we do?

Only with how you have it implemented in this branch. Diff'd against master, I could see something (roughly) like:

```diff
diff --git a/blocks/library/embed/index.js b/blocks/library/embed/index.js
index ca0762e24..f60f8da5c 100644
--- a/blocks/library/embed/index.js
+++ b/blocks/library/embed/index.js
@@ -26,6 +26,23 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
 // These embeds do not work in sandboxes
 const HOSTS_NO_PREVIEWS = [ 'facebook.com' ];
 
+const matchesPatterns = ( url, patterns ) => {
+	return patterns.some( ( pattern ) => {
+		return url.match( new RegExp( pattern, 'i' ) );
+	} );
+};
+
+const findBlock = ( url ) => {
+	for ( const blocks of [ common, others ] ) {
+		for ( const block of blocks ) {
+			if ( matchesPatterns( url, block.patterns ) ) {
+				return block.name;
+			}
+		}
+	}
+	return 'core/embed';
+};
+
 function getEmbedBlockSettings( { title, description, icon, category = 'embed', transforms, keywords = [] } ) {
 	return {
 		title,
@@ -359,6 +376,7 @@ export const others = [
 			title: 'Animoto',
 			icon: 'embed-video',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/cloudup',
@@ -366,6 +384,7 @@ export const others = [
 			title: 'Cloudup',
 			icon: 'embed-post',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/collegehumor',
@@ -373,6 +392,7 @@ export const others = [
 			title: 'CollegeHumor',
 			icon: 'embed-video',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/dailymotion',
@@ -380,6 +400,7 @@ export const others = [
 			title: 'Dailymotion',
 			icon: 'embed-video',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/funnyordie',
@@ -387,6 +408,7 @@ export const others = [
 			title: 'Funny or Die',
 			icon: 'embed-video',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/hulu',
@@ -394,6 +416,7 @@ export const others = [
 			title: 'Hulu',
 			icon: 'embed-video',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/imgur',
@@ -401,6 +424,7 @@ export const others = [
 			title: 'Imgur',
 			icon: 'embed-photo',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/issuu',
@@ -408,6 +432,7 @@ export const others = [
 			title: 'Issuu',
 			icon: 'embed-post',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/kickstarter',
@@ -415,6 +440,7 @@ export const others = [
 			title: 'Kickstarter',
 			icon: 'embed-post',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/meetup-com',
@@ -422,6 +448,7 @@ export const others = [
 			title: 'Meetup.com',
 			icon: 'embed-post',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/mixcloud',
@@ -430,6 +457,7 @@ export const others = [
 			icon: 'embed-audio',
 			keywords: [ __( 'music' ), __( 'audio' ) ],
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/photobucket',
@@ -437,6 +465,7 @@ export const others = [
 			title: 'Photobucket',
 			icon: 'embed-photo',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/polldaddy',
@@ -444,6 +473,7 @@ export const others = [
 			title: 'Polldaddy',
 			icon: 'embed-post',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/reddit',
@@ -451,6 +481,7 @@ export const others = [
 			title: 'Reddit',
 			icon: 'embed-post',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/reverbnation',
@@ -458,6 +489,7 @@ export const others = [
 			title: 'ReverbNation',
 			icon: 'embed-audio',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/screencast',
@@ -465,6 +497,7 @@ export const others = [
 			title: 'Screencast',
 			icon: 'embed-video',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/scribd',
@@ -472,6 +505,7 @@ export const others = [
 			title: 'Scribd',
 			icon: 'embed-post',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/slideshare',
@@ -479,6 +513,7 @@ export const others = [
 			title: 'Slideshare',
 			icon: 'embed-post',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/smugmug',
@@ -486,6 +521,7 @@ export const others = [
 			title: 'SmugMug',
 			icon: 'embed-photo',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/speaker',
@@ -493,6 +529,7 @@ export const others = [
 			title: 'Speaker',
 			icon: 'embed-audio',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/ted',
@@ -500,6 +537,7 @@ export const others = [
 			title: 'TED',
 			icon: 'embed-video',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/tumblr',
@@ -507,6 +545,7 @@ export const others = [
 			title: 'Tumblr',
 			icon: 'embed-post',
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/videopress',
@@ -515,6 +554,7 @@ export const others = [
 			icon: 'embed-video',
 			keywords: [ __( 'video' ) ],
 		} ),
+		patterns: [],
 	},
 	{
 		name: 'core-embed/wordpress-tv',
@@ -522,5 +562,6 @@ export const others = [
 			title: 'WordPress.tv',
 			icon: 'embed-video',
 		} ),
+		patterns: [],
 	},
 ];
```

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -110,20 +132,45 @@ function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, k
 
+				// If we don't have any URL patterns, or we do and the URL doesn't match,
+				// then we should look for a block that has a matching URL pattern.
+				if ( ! patterns || ( patterns && ! matchesPatterns( url, patterns ) ) ) {
```
New commit pushed up that removes this Python influence :)

iseulde commented:
Are there plans to allow plugins to extend the embeds available just like WordPress core does now? Would we be better off with registering these block definitions on the server? Sorry if I'm missing some background here.

notnownikki commented:
@aduth this should have addressed the last couple of review points. `patterns` is back in the `other` and `common` lists, and they're regexes not strings.

aduth commented on blocks/library/embed/index.js:
```diff
@@ -26,7 +27,25 @@ import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+	return 'core/embed';
+};
+
+function getEmbedBlockSettings( { title, description, icon, category = 'embed', transforms, keywords = [], patterns = [] } ) {
```
We're no longer passing `patterns` here, so all of the logic related to this, specifically on line 130, will not be triggered. In practice, it's not an issue because `findBlock` will get triggered and find the correct block anyways, but seems we might not need this as a parameter.

notnownikki replied:
ah yes, removing it now!

notnownikki committed:
Transform into the correct embed block based on URL patterns

notnownikki commented:
Rebased and squished commits. Took out the `patterns` parameter and related code from the block settings.

aduth commented on core-blocks/embed/index.js:
```diff
@@ -28,6 +29,24 @@ import './editor.scss';
+const wpEmbedAPI = memoize( ( url ) => wp.apiRequest( { path: `/oembed/1.0/proxy?${ stringify( { url } ) }` } ) );
+
+const matchesPatterns = ( url, patterns ) => {
+	return patterns.some( ( pattern ) => {
```
Noting that this line can error:

1. Paste `https://wordpress.org/gutenberg/handbook` into a new paragraph block

```
Uncaught TypeError: Cannot read property 'some' of undefined
```

Simple solution may just be to add a default parameter value for `patterns`:

```js
const matchesPatterns = ( url, patterns = [] ) => {
```

aduth committed:
Embed: Add default patterns value for matchesPatterns

aduth commented:
I added the default argument value for `patterns` in fc86d74 , but I'm still having issues with the same example: It transforms into an embed block, but never embeds anything, even when I continue with pressing "Embed" explicitly. If it's not embeddable, should it not be communicating this to me? Has this regressed?

notnownikki commented:
Yes this is a regression. I'll look at it tomorrow, and see about adding
some puppeteer tests too.

On Wed, 25 Apr 2018, 21:05 Andrew Duthie, <notifications@github.com> wrote:

> I added the default argument value for patterns in fc86d74
> <https://github.com/WordPress/gutenberg/commit/fc86d74a9ea8100a0e32b4821c583ab07fa03f80>
> , but I'm still having issues with the same example: It transforms into an
> embed block, but never embeds anything, even when I continue with pressing
> "Embed" explicitly. If it's not embeddable, should it not be communicating
> this to me? Has this regressed?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/5315#issuecomment-384416748>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADKSssuOKgiEiV_69SK1-JZsIEe0xRwPks5tsNcagaJpZM4SXPeP>
> .
>

notnownikki committed:
Don't error if the url is undefined

notnownikki commented:
@aduth I took a look at this and can't reproduce the problem. I tried embeddable urls, unembeddable urls, and urls that don't resolve to a valid host... I tried pasting and entering the url manually, and I get the error message in the correct scenarios. Can you take a look at this gif and let me know if I've missed out the test you did? (the first URL, http://notnownikki.com/sameties/ is the pasted one)

![embed](https://user-images.githubusercontent.com/3314354/39291900-5b057bf6-492c-11e8-8479-ce90fb4d88ab.gif)

aduth commented:
Did you try with this link specifically: `https://wordpress.org/gutenberg/handbook`

I get a “Cannot embed” message with the URL from your comment, but nothing for this one.

notnownikki committed:
Correct logic from when the support for 'result' was removed

aduth commented on core-blocks/embed/index.js:
```diff
@@ -109,27 +128,53 @@ function getEmbedBlockSettings( { title, description, icon, category = 'embed',
 				const { url } = this.props.attributes;
 				const { setAttributes } = this.props;
 
+				if ( undefined === url ) {
```
Is the condition in `componentWillMount` redundant now? For that matter, is the `setState` redundant as well?

notnownikki replied:
Ah, good catch! Removed!

notnownikki committed:
Removed unneeded conditionals

aduth commented:
Actually, I see this on master as well. Maybe a regression of #5966 (cc @iseulde @mtias).

iseulde commented:
Looking

notnownikki commented:
I'm seeing the problem on master, but not on the non-rebased version of this branch

iseulde commented:
I think it has to do with the priority that has been removed from paragraph.

aduth reviewed (approved):
Needs conflicts resolved, but otherwise LGTM 👍

notnownikki committed:
Merge branch 'master' into update/embed-switch-to-correct-embed-block
